### PR TITLE
Block server type update when server is assigned DSes

### LINKED ISSUE #4113: TO: Can Update a Server's Type after its been assigned to Delivery Service
## I'm submitting a ...

- [X] bug report
- [ ] new feature / enhancement request
- [ ] improvement request (usability, performance, tech debt, etc.)
- [ ] other 

## Traffic Control components affected ...

- [ ] CDN in a Box
- [ ] Documentation
- [ ] Grove
- [ ] Traffic Control Client
- [ ] Traffic Monitor
- [X] Traffic Ops
- [ ] Traffic Ops ORT
- [ ] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [ ] Traffic Vault
- [ ] unknown

## Current behavior:

Delivery Service Required Capability is created before any Servers are assigned to a Delivery Service. 
Making a  `POST` Request to `https://{{TO_BASE_URL}}/api/{{api_version}}/deliveryserviceserver` with a server of  type other than `EDGE` is allowed in the API.
Updating that same server: `PUT` `https://{{TO_BASE_URL}}/api/{{api_version}}/servers/{id}/` to Type `EDGE` is allowed. Even though that server does not have the required capability associated to it.
## Expected / new behavior:

1. Should not be allowed to link a server of ANY type to a delivery service, ONLY Edge and Origin
2. Should not be allowed to update a server's type when linked to a delivery service.

### PR DESCRIPTION
**Pull request by mhoppa**
## What does this PR (Pull Request) do?

- [x] This PR fixes #4113 

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

Run API tests or manually attempt to change a servers type that is assigned to DSes and that is not assigned to DSes. The latter should pass but the first should pass.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: traffic_ops/testing/api/v14/servers_test.go [hunk 1]
```diff
@@ -87,6 +87,41 @@ func UpdateTestServers(t *testing.T) {
...
+	if len(dses) < 1 {
+		t.Fatal("GET DeliveryServices returned no dses, must have at least 1 to test invalid type server update")
+	}
+
+	serverTypes, _, err := TOSession.GetTypes("server")
+	if err != nil {
+		t.Fatalf("cannot GET Server Types: %v", err)
+	}
+	if len(serverTypes) < 2 {
+		t.Fatal("GET Server Types returned no types, must have at least 2 to test invalid type server update")
```
Comment (ocket8888) [inline traffic_ops/testing/api/v14/servers_test.go]: It's possible that it returned 1 type, not no types

File: traffic_ops/traffic_ops_golang/server/servers.go [hunk 2]
```diff
@@ -36,7 +39,7 @@ import (
 
 	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/deliveryservice"
 	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/tenant"
-	"github.com/go-ozzo/ozzo-validation"
+	validation "github.com/go-ozzo/ozzo-validation"
```
Comment (ocket8888) [inline traffic_ops/traffic_ops_golang/server/servers.go]: why?

File: traffic_ops/traffic_ops_golang/server/servers.go [hunk 3]
```diff
@@ -322,6 +325,27 @@ func (sv *TOServer) Update() (error, error, int) {
...
+	}
+
+	// If type is changing ensure it isnt assigned to any DSes
+	if typeID != *sv.TypeID {
+		dsIDs := []int64{}
+		if err := sv.APIInfo().Tx.QueryRowx("SELECT ARRAY(SELECT deliveryservice FROM deliveryservice_server WHERE server = $1)", sv.ID).Scan(pq.Array(&dsIDs)); err != nil && err != sql.ErrNoRows {
+			return nil, fmt.Errorf("getting server assigned delivery services: %v", err), http.StatusInternalServerError
+		}
+		if len(dsIDs) != 0 {
+			return errors.New("server type can not be updated when it is currently assigned to delivery services"), nil, http.StatusBadRequest
```
Comment (ocket8888) [inline traffic_ops/traffic_ops_golang/server/servers.go]: Personally I like [`409 Conflict`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409) for things like this. Do with that what you will.

File: traffic_ops/testing/api/v14/servers_test.go (hunk 1 above)
Comment (mhoppa) [reply traffic_ops/testing/api/v14/servers_test.go]: good catch

File: traffic_ops/traffic_ops_golang/server/servers.go (hunk 2 above)
Comment (mhoppa) [reply traffic_ops/traffic_ops_golang/server/servers.go]: just matched how its being called in code but 🤷‍♂ ill change it back

Comment (ocket8888) [reply traffic_ops/traffic_ops_golang/server/servers.go]: I've seen that in few places, but the package is named `validation` already, so idk why. Maybe it used to be necessary when the package name didn't match the terminating directory name?

Comment (rob05c) [reply traffic_ops/traffic_ops_golang/server/servers.go]: +1 on removing the alias. Yeah, we have a bunch of places that do it, I'm guessing whoever added the library first didn't realize Go follows the package name, not the directory; and everyone else just copied it. If it's an idiom, IMO it's an antipattern and we should remove them.

**Pull request merged**